feat: Implement JSON Schema Validation for User API Responses

AmpasTheodoros opened the pull request:
Problem Solved: The Cal.com API v2 was exhibiting inconsistent behavior with response bodies changing unpredictably. This inconsistency was causing errors in applications relying on the API, leading to disruptions in functionality and user experience. The issue stemmed from a lack of a fixed and documented schema for API responses

Solution Implemented:

- Defined a JSON schema to establish a consistent structure for user-related API responses
- Integrated the Ajv library into the UsersService to validate responses against the schema
- Added validation logic in the getByUsernames method to ensure each user response adheres to the schema, throwing an error if it does not

Trade-offs or Limitations:

- The current implementation focuses on user-related responses. Extending schema validation to other modules may be necessary for comprehensive consistency
- Future changes to the API response structure will require updates to the schema and potentially the validation logic
- A versioning strategy and deprecation policy should be considered to manage future changes and maintain backward compatibility

AmpasTheodoros committed:
Implement JSON Schema Validation for User API Responses

Problem Solved: The Cal.com API v2 was exhibiting inconsistent behavior with response bodies changing unpredictably. This inconsistency was causing errors in applications relying on the API, leading to disruptions in functionality and user experience. The issue stemmed from a lack of a fixed and documented schema for API responses.

Solution Implemented:

Defined a JSON schema to establish a consistent structure for user-related API responses.
Integrated the Ajv library into the UsersService to validate responses against the schema.
Added validation logic in the getByUsernames method to ensure each user response adheres to the schema, throwing an error if it does not.
Trade-offs or Limitations:

The current implementation focuses on user-related responses. Extending schema validation to other modules may be necessary for comprehensive consistency.
Future changes to the API response structure will require updates to the schema and potentially the validation logic.
A versioning strategy and deprecation policy should be considered to manage future changes and maintain backward compatibility.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=17731) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=17731) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=17731) it.</sub>

Praashh commented:
Hey @AmpasTheodoros please sign CLA

Praashh commented on apps/api/v2/src/modules/users/services/users.service.ts:
```diff
@@ -19,7 +23,14 @@ export class UsersService {
+    // Validate each user response
+    usersFiltered.forEach((user) => {
+      if (!validateUserResponse(user)) {
+        throw new Error("Invalid user response schema");
```
```suggestion
  throw new Error(`Invalid user response schema for user: ${user.username}`);
```

we can throw the error for the user.

AmpasTheodoros commented:
Hey @Praashh  thanks for fixing my title I just signed CLA

Praashh reviewed (dismissed):
LGTM!!

keithwillcode reviewed (changes requested):
Looks like API v2 tests are failing. Can you please take a look?

AmpasTheodoros committed:
fix: Resolve Ajv Schema Compilation Error in Users Service

Problem Solved:
The API test suite was failing due to an Ajv schema compilation error in the Users Service. The error "schema must be object or boolean" was occurring during module initialization, blocking the execution of e2e tests and affecting multiple dependent modules including users, auth, and organizations.

Solution Implemented:
- Added proper JSON module type assertion for schema import
- Applied type assertion when compiling schema with Ajv to ensure correct object handling
- Maintained the existing schema structure while fixing the import mechanism

Trade-offs or Limitations:
- Using type assertions is a temporary solution and might need revisiting for stricter type safety
- The current fix focuses on compilation issues rather than schema validation logic
- Future schema updates should consider maintaining backward compatibility

AmpasTheodoros committed:
refactor: Remove unnecessary type assertion for JSON schema

The type assertion `as Record<string, unknown>` was redundant since:
- resolveJsonModule is enabled in tsconfig
- userResponseSchema.json is a valid JSON Schema
- TypeScript can properly infer the schema type

This change improves type safety and code cleanliness while maintaining the same functionality.

AmpasTheodoros committed:
fix: Correct calibration calculation in round-robin scheduling

The calibration calculation in getHostsWithCalibration was not accurately
representing the booking rate when new hosts were added. Changes made:

1. Include all historical bookings before host addition, not just other users'
2. Consider all previous hosts when calculating the denominator
3. Simplify division-by-zero check

This fixes the test case where calibration was expected to be 2.5 but was
returning 0, making the round-robin scheduling more fair for newly added hosts

anikdhabal commented:
Close due to inactivity
